main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala

```diff
@@ -0,0 +1,1069 @@
+             | SemanticOp.Int32Op.Eq | SemanticOp.Int32Op.Neq
+             | SemanticOp.Int64Op.Eq | SemanticOp.Int64Op.Neq =>
+          val (tpe1, eff1) = visitExp(exp1)
+          val (tpe2, eff2) = visitExp(exp2)
+          c.unifyTypeM(tpe1, tpe2, loc)
+          c.unifyTypeM(tvar, Type.Bool, loc)
+          val resTpe = tvar
+          val resEff = Type.mkUnion(eff1, eff2, loc)
+          (resTpe, resEff)
+
+        case SemanticOp.CharOp.Lt | SemanticOp.CharOp.Le | SemanticOp.CharOp.Gt | SemanticOp.CharOp.Ge
+             | SemanticOp.Float32Op.Lt | SemanticOp.Float32Op.Le | SemanticOp.Float32Op.Gt | SemanticOp.Float32Op.Ge
+             | SemanticOp.Float64Op.Lt | SemanticOp.Float64Op.Le | SemanticOp.Float64Op.Gt | SemanticOp.Float64Op.Ge
+             | SemanticOp.Int8Op.Lt | SemanticOp.Int8Op.Le | SemanticOp.Int8Op.Gt | SemanticOp.Int8Op.Ge
+             | SemanticOp.Int16Op.Lt | SemanticOp.Int16Op.Le | SemanticOp.Int16Op.Gt | SemanticOp.Int16Op.Ge
+             | SemanticOp.Int32Op.Lt | SemanticOp.Int32Op.Le | SemanticOp.Int32Op.Gt | SemanticOp.Int32Op.Ge
+             | SemanticOp.Int64Op.Lt | SemanticOp.Int64Op.Le | SemanticOp.Int64Op.Gt | SemanticOp.Int64Op.Ge =>
+          val (tpe1, eff1) = visitExp(exp1)
+          val (tpe2, eff2) = visitExp(exp2)
+          c.unifyTypeM(tpe1, tpe2, loc)
+          c.unifyTypeM(tvar, Type.Bool, loc)
+          val resTpe = tvar
+          val resEff = Type.mkUnion(eff1, eff2, loc)
+          (resTpe, resEff)
+
+        case SemanticOp.StringOp.Concat =>
+          val (tpe1, eff1) = visitExp(exp1)
+          val (tpe2, eff2) = visitExp(exp2)
+          c.expectTypeM(expected = Type.Str, actual = tpe1, exp1.loc)
+          c.expectTypeM(expected = Type.Str, actual = tpe2, exp2.loc)
+          c.unifyTypeM(tvar, Type.Str, loc)
+          val resTpe = tvar
+          val resEff = Type.mkUnion(eff1, eff2, loc)
+          (resTpe, resEff)
+
+        case _ => throw InternalCompilerException(s"Unexpected binary operator: '$sop'.", loc)
+      }
+
+      case Expr.IfThenElse(exp1, exp2, exp3, loc) =>
+        val (tpe1, eff1) = visitExp(exp1)
+        val (tpe2, eff2) = visitExp(exp2)
+        val (tpe3, eff3) = visitExp(exp3)
+        c.expectTypeM(expected = Type.Bool, actual = tpe1, exp1.loc)
+        c.unifyTypeM(tpe2, tpe3, loc)
+        val resTpe = tpe3
+        val resEff = Type.mkUnion(eff1, eff2, eff3, loc)
+        (resTpe, resEff)
+
+      case Expr.Stm(exp1, exp2, loc) =>
+        val (_, eff1) = visitExp(exp1)
+        val (tpe2, eff2) = visitExp(exp2)
+        val resTpe = tpe2
+        val resEff = Type.mkUnion(eff1, eff2, loc)
+        (resTpe, resEff)
+
+      case Expr.Discard(exp, _) =>
+        val (_, eff) = visitExp(exp)
+        val resTpe = Type.Unit
+        val resEff = eff
+        (resTpe, resEff)
+
+      case Expr.Let(sym, _, exp1, exp2, loc) =>
+        val (tpe1, eff1) = visitExp(exp1)
+        c.unifyTypeM(sym.tvar, tpe1, loc)
```

Comment:
let uses `loc` and letrec uses `exp1.loc`. which is best?

Reply:
exp1.loc is more precise. fixed